src/util/retryable.ts

```diff
@@ -0,0 +1,30 @@
+export type RetryOptions = {
+  maxAttempts: number;
+  maxBackoff: number;
+  shouldRetry?: (error: any) => boolean;
+  attempt?: number;
+};
+
+export const withRetries = async <T>(
+  fn: () => Promise<T>,
+  {
+    maxAttempts,
+    maxBackoff,
+    shouldRetry = () => true,
+    attempt = 0,
+  }: RetryOptions,
+): Promise<T> => {
+  const backoffMs = Math.min(Math.random() * 2 ** attempt, maxBackoff) * 1_000;
```

Comment:
this delays the _first_ request, right? at attempt 0, it's 2^0=1, so we're delaying by 0-1,000 ms.

should either special-case attempt 0 or pick a formula that evaluates to 0 naturally on attempt 0. maybe worth a unit test to validate/prevent regression.

Reply:
It tries to run `fn()` before any retries, so I think we are good there.

How _would_ we test? 🤔 

If we added an argument for a fn that calculates the delay, then we could test with a delay that doesn't use randomness and could depend on. Otherwise, I think we'd have to run a base test several times and check that the average execution time is below an amount.

I do think that customizable retries would be a great follow-up. Including how to pick which responses are retried and the algorithm for delays.

Reply:
👍 

Stubbing `math.random()` and `setTimeout` should work for testing a situation like this.

Reply:
I can add tests. NDB.